diff --git a/demo.py b/demo.py
index 604b5a1..c991ef1 100644
--- a/demo.py
+++ b/demo.py
@@ -11,6 +11,8 @@ import open3d as o3d
 from core.deep_global_registration import DeepGlobalRegistration
 from config import get_config
 
+import pyplotlib.pyplot as plt
+
 BASE_URL = "http://node2.chrischoy.org/data/"
 DOWNLOAD_LIST = [
     (BASE_URL + "datasets/registration/", "redkitchen_000.ply"),
@@ -40,9 +42,16 @@ if __name__ == '__main__':
   dgr = DeepGlobalRegistration(config)
   T01 = dgr.register(pcd0, pcd1)
 
-  o3d.visualization.draw_geometries([pcd0, pcd1])
+  vis = o3d.visualization.Visualizer()
+  vis.add_geometries([pcd0, pcd1])
+  vis.capture_screen_float_buffer(False)
+  plg.imsave("./before.png")
 
   pcd0.transform(T01)
   print(T01)
 
   o3d.visualization.draw_geometries([pcd0, pcd1])
+  vis = o3d.visualization.Visualizer()
+  vis.add_geometries([pcd0, pcd1])
+  vis.capture_screen_float_buffer(False)
+  plg.imsave("./after.png")
\ No newline at end of file
diff --git a/exec.sh b/exec.sh
index 84d9e31..e3bf4b0 100755
--- a/exec.sh
+++ b/exec.sh
@@ -1,5 +1,5 @@
 #!/bin/sh
-APPDIR=`dirname $0`
-pip install -r $APPDIR/requirements.txt --user
+pip install --upgrade pip>=20.3 --user
+pip install -r /repo/requirements.txt --user
 
-python -u demo.py
+python -u /repo/demo.py
